### ISSUE TITLE
Suggestion for connecting in an LSTM or Differentiable Neural Computer?

### ISSUE THREAD

**Issue by SheldonCurtiss** [JimOhman/model-based-rl#11]
How would you suggest to rig up an LSTM or Differentiable Neural Computer into this most effectively?
Is this even possible?

**Comment by JimOhman** [comment]
Hmm, putting an LSTM, say in the representation network, would include storing the hidden states with each step in the game.history and reattaching these before the forward pass of the learner (this should be straight forward). Then you likely need to sample a longer sequence than num_unroll_steps such that you can dedicate a part of that for a "burn in" to recover from stale hidden states (this change shouldn't be too bad either). You'll also likely benefit from changing the way priorities are calculated to include the sequences and not just the MCTS. The ideas relate to how this paper (https://openreview.net/pdf?id=r1lyTjAqYX) deals with LSTM in a distributed setting.

You could create a class for an LSTMActor and LSTMLearner that inherits from Actor and Learner and overwrites the methods that are necessary. So I think it's possible to add a LSTM core for the representation network, without too much work.

Feel free to contribute these additions if you want.

**Comment by SheldonCurtiss** [comment]
Thank you for your response.

I think I follow and agree with you in regards to "sample a longer sequence than num_unroll_steps such that you can dedicate a part of that for a "burn in" to recover from stale hidden states"
I'm currently also investigating how a few of the gym examples utilize ram.
Most of the various implementations I've used self-play on GPU is incredibly slow and not worth it which is another thing I plan to investigate.

Do you have any opinions regarding "Differentiable Neural Computer"? Muzero seems Niche but DNC seems hyper niche. It's difficult to gather useful information on.
Also what is your opinion regarding Neuroevolution? I saw a gpu implementation the other day that's peaked my interest. I would think in theory Muzero would learn faster though.

**Comment by JimOhman** [comment]
I have a feeling DNC's could be used with the dynamics head and in conjunction to an LSTM for the representation head. However, I don't know enough knowledge about DNC's at this moment to say for sure, but I'm looking into it.

Neuroevolution is certainly interesting, but I think that gradient descent and evolutionary algorithms should be used together, such that gradient descent fine-tunes and the evolution is some sort of meta-learner. So it might be possible to combine the two for MuZero, but I have a feeling the compute requirements would be pretty high.
